src/reconcile/rec_write.c

```diff
@@ -573,16 +572,29 @@ __rec_write_check_complete(WT_SESSION_IMPL *session, WT_RECONCILE *r)
 		return (EBUSY);
 
 	/*
-	 * If we are doing update/restore based eviction, confirm part of the
-	 * page is being discarded, or at least 10% of the updates won't have
-	 * to be re-instantiated. Otherwise, it isn't progress, don't bother.
+	 * If doing update/restore based eviction, see if rewriting the page in
+	 * memory is worth the effort.
 	 */
 	if (F_ISSET(r, WT_EVICT_UPDATE_RESTORE)) {
-		for (bnd = r->bnd, i = 0; i < r->bnd_entries; ++bnd, ++i)
+		/* If discarding a disk-page size chunk, do it. */
+		for (bnd = r->bnd, i = 0; i < r->bnd_next; ++bnd, ++i)
 			if (bnd->supd == NULL)
-				break;
-		if (i == r->bnd_entries &&
-		    r->update_cnt / 10 >= r->update_skip_cnt)
+				return (0);
+
+		/*
+		 * Switch to the lookaside table if we can: it's more effective
+		 * than rewriting a page in memory because it implies eviction.
+		 */
+		if (r->update_mem_skipped == 0)
+			return (EBUSY);
+
+		/*
+		 * Don't rewrite pages where we're not going to get back enough
+		 * memory to care. There's no empirical evidence the 10KB limit
+		 * is a good configuration, but it should keep us from wasting
+		 * time on tiny pages and pages with only a few updates.
+		 */
+		if (r->update_mem - r->update_mem_skipped < 10 * WT_KILOBYTE)
```

Comment:
I don't much like the 10KB limit, either, but in heavy insert workloads under cache stress, I saw lots of small pages being evicted, pages only a few K in size, with only a few update chains.

```c
if (r->update_mem_skipped > 9 * r->update_mem / 10)
    return (EBUSY);
```

This won't catch pages with only a couple of update chains we had to skip. Imagine one update chain of 100 bytes we had to skip, `100 > (9 * 100 / 10) == true`.

We don't want rewrite pages in memory if we're not going to get anything back. We could drop the limit to 1-2KB? Is it worth doing any work at all to recover 2KB?